source/magnetic_induction/assembly_advection_matrix.cc

```diff
@@ -0,0 +1,275 @@
+      velocity->old_solution,
+      scratch.old_velocity_gradients);
+
+    scratch.velocity_fe_values[vector_extractor].get_function_gradients(
+      velocity->old_old_solution,
+      scratch.old_old_velocity_gradients);
+
+    scratch.velocity_fe_values[vector_extractor].get_function_divergences(
+      velocity->old_solution,
+      scratch.old_velocity_divergences);
+
+    scratch.velocity_fe_values[vector_extractor].get_function_divergences(
+      velocity->old_old_solution,
+      scratch.old_old_velocity_divergences);
+
+    for (unsigned int i = 0; i < scratch.n_q_points; ++i)
+    {
+      scratch.velocity_values[i] =
+        eta[0] * scratch.old_velocity_values[i] +
+        eta[1] * scratch.old_old_velocity_values[i];
+
+      scratch.velocity_gradients[i] =
+        eta[0] * scratch.old_velocity_gradients[i] +
+        eta[1] * scratch.old_old_velocity_gradients[i];
+
+      scratch.velocity_divergences[i] =
+        eta[0] * scratch.old_velocity_divergences[i] +
+        eta[1] * scratch.old_old_velocity_divergences[i];
+    }
+  }
+  else if (ptr_velocity_function != nullptr)
+  {
+    ptr_velocity_function->value_list(
+      scratch.magnetic_field_fe_values.get_quadrature_points(),
+      scratch.velocity_values);
+
+    ptr_velocity_function->gradient_list(
+      scratch.magnetic_field_fe_values.get_quadrature_points(),
+      scratch.velocity_gradients);
+
+    for (unsigned int i = 0; i < scratch.n_q_points; ++i)
+      scratch.velocity_divergences[i] = trace(scratch.velocity_gradients[i]);
+  }
+  else
+  {
+    ZeroTensorFunction<1,dim>().value_list(
+      scratch.magnetic_field_fe_values.get_quadrature_points(),
+      scratch.velocity_values);
+
+    ZeroTensorFunction<2,dim>().value_list(
+      scratch.magnetic_field_fe_values.get_quadrature_points(),
+      scratch.velocity_gradients);
+
+    ZeroFunction<dim>().value_list(
+      scratch.magnetic_field_fe_values.get_quadrature_points(),
+      scratch.velocity_divergences);
+  }
+
+  // Loop over quadrature points
+  for (unsigned int q = 0; q < scratch.n_q_points; ++q)
+  {
+    // Extract test function values at the quadrature points
+    for (unsigned int i = 0; i < scratch.dofs_per_cell; ++i)
+    {
+      scratch.phi[i]      = scratch.magnetic_field_fe_values[vector_extractor].value(i, q);
+      scratch.grad_phi[i] = scratch.magnetic_field_fe_values[vector_extractor].gradient(i, q);
+      scratch.div_phi[i]  = scratch.magnetic_field_fe_values[vector_extractor].divergence(i,q);
+    }
+
+    // Loop over local degrees of freedom
+    for (unsigned int i = 0; i < scratch.dofs_per_cell; ++i)
+      for (unsigned int j = 0; j < scratch.dofs_per_cell; ++j)
+        data.local_matrix(i,j) +=
+          scratch.phi[i] *
+          (scratch.velocity_values[q] *
+           scratch.div_phi[j]
+           -
+           scratch.phi[j] *
+           scratch.velocity_divergences[q]
+           +
+           scratch.velocity_gradients[q] *
+           scratch.phi[j]
```

Comment:
I would move this term to the right-hand side and interpret it as a _source term_.